streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -103,36 +104,46 @@ public void validatePartitionExists(final TopicPartition topicPartition, final S
     public void register(final StateRestorer restorer) {
         restorer.setGlobalRestoreListener(stateRestoreListener);
         stateRestorers.put(restorer.partition(), restorer);
+        // if it has already been initialized which can happen if there was a lock exception
+        // so some tasks may have been restored
+        if (initialized) {
+            addedPostInitialization.put(restorer.partition(), restorer);
+        }
     }
 
 
     public Collection<TopicPartition> restore() {
-        final List<TopicPartition> completed = new ArrayList<>();
         if (!initialized) {
-            completed.addAll(initialize());
+            needsRestoring.clear();
+            initialize(stateRestorers);
+        }
+
+        if (!addedPostInitialization.isEmpty()) {
+            initialize(addedPostInitialization);
+            addedPostInitialization.clear();
         }
 
         if (needsRestoring.isEmpty()) {
             consumer.assign(Collections.<TopicPartition>emptyList());
-            return completed;
+            return completed();
         }
 
         final Set<TopicPartition> partitions = new HashSet<>(needsRestoring.keySet());
         final ConsumerRecords<byte[], byte[]> allRecords = consumer.poll(10);
         for (final TopicPartition partition : partitions) {
-            if (restorePartition(allRecords, partition)) {
-                completed.add(partition);
-            }
+            restorePartition(allRecords, partition);
```

Comment:
Does `restorePartition` still need to return a boolean then?

Reply:
nope